src/Streamly/Internal/Unicode/Utf8/Type.hs

```diff
@@ -0,0 +1,474 @@
+-- Stability   : experimental
+-- Portability : GHC
+--
+module Streamly.Internal.Unicode.Utf8.Type
+    (
+    -- * Type
+      Utf8
+    , toArray
+
+    -- * Creation and elimination
+    , empty
+    , singleton
+    , toStream
+    , fromStream
+    , pack
+    , unpack
+
+    -- * Basic interface
+    , cons
+    , snoc
+    , append
+    , uncons
+    , unsnoc
+    , head
+    , last
+    , tail
+    , init
+    , null
+
+    , isSingleton
+    , length
+    , compareLength
+
+    -- * Folds
+    , write
+
+    -- * Unfolds
+    , read
+    )
+where
+
+
+#include "inline.hs"
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.DeepSeq (NFData)
+import Control.Monad.IO.Class (MonadIO)
+import Data.Bits (shiftR, (.&.))
+import Data.Char (ord)
+import Data.String (IsString(..))
+import Data.Word (Word8)
+import GHC.Base (assert)
+import GHC.IO.Encoding.Failure (isSurrogate)
+import Streamly.Internal.Data.Array.Foreign.Type (Array)
+import Streamly.Internal.Data.Fold (Fold)
+import Streamly.Internal.Data.Stream.IsStream (SerialT)
+import Streamly.Internal.Data.Unfold (Unfold)
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Streamly.Internal.Data.Array.Foreign as Array
+import qualified Streamly.Internal.Data.Array.Foreign.Type as Array
+import qualified Streamly.Internal.Data.Array.Foreign.Mut.Type as MArray
+import qualified Streamly.Internal.Data.Fold.Type as Fold
+import qualified Streamly.Internal.Data.Stream.IsStream as Stream
+import qualified Streamly.Internal.Unicode.Stream as Unicode
+
+import Prelude hiding
+    ( all
+    , any
+    , break
+    , concat
+    , concatMap
+    , drop
+    , dropWhile
+    , elem
+    , filter
+    , foldl
+    , foldl1
+    , foldr
+    , foldr1
+    , head
+    , init
+    , last
+    , length
+    , lines
+    , map
+    , maximum
+    , minimum
+    , null
+    , read
+    , replicate
+    , reverse
+    , scanl
+    , scanl1
+    , scanr
+    , scanr1
+    , span
+    , splitAt
+    , tail
+    , take
+    , takeWhile
+    , unlines
+    , unwords
+    , words
+    , zip
+    , zipWith
+    )
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+-- >>> import qualified Streamly.Internal.Unicode.Utf8 as Utf8
+
+--------------------------------------------------------------------------------
+-- Type
+--------------------------------------------------------------------------------
+
+-- | A space efficient, packed, unboxed Unicode container.
+newtype Utf8 =
+    Utf8 (Array Word8)
+    deriving (NFData)
+
+empty :: Utf8
+empty = Utf8 Array.nil
+
+--------------------------------------------------------------------------------
+-- Helpers
+--------------------------------------------------------------------------------
+
+{-# INLINE toStream #-}
+toStream :: Utf8 -> SerialT IO Char
+toStream = Unicode.decodeUtf8' . Array.toStream . toArray
```

Comment:
Why IO? Can't we use Identity or `Monad m` here? It will require MonadIO constraint everywhere which is not good.

Reply:
Ah, I've written IO keeping mutable arrays in mind. It can be removed for immutable stuff.